janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/HasStepFolder.java

```diff
@@ -168,13 +167,13 @@ static void foldInHasContainer(final HasStepFolder janusgraphStep, final Travers
                 ((OrStep<?>) currentStep).getLocalChildren().forEach(t ->localFoldInHasContainer(janusgraphStep, t.getStartStep(), t, rootTraversal));
                 currentStep.getLabels().forEach(janusgraphStep::addLabel);
                 traversal.removeStep(currentStep);
-            } else if (currentStep instanceof HasContainerHolder){
-                final Iterable<HasContainer> containers = ((HasContainerHolder) currentStep).getHasContainers().stream().map(JanusGraphPredicateUtils::convert).collect(Collectors.toList());
-                if  (validFoldInHasContainer(currentStep, true)) {
-                    janusgraphStep.addAll(containers);
-                    currentStep.getLabels().forEach(janusgraphStep::addLabel);
-                    traversal.removeStep(currentStep);
-                }
+            } else if (currentStep instanceof HasContainerHolder && validFoldInHasContainer(currentStep, true)){
+                final HasContainerHolder currentHasContainerHolder = (HasContainerHolder) currentStep;
+                final Iterable<HasContainer> containers = () -> currentHasContainerHolder.getHasContainers()
+                    .stream().map(JanusGraphPredicateUtils::convert).iterator();
```

Comment:
(somewhat) nitpick: I remember you once discovered that unnecessary usage of `stream()` can cause computation overhead. You could easily avoid streaming here and replace it by `currentHasContainerHolder.getHasContainers().forEach(...)`.
`JanusGraphStep` only has an `addAll()` method, but no `add()`. Maybe it would make sense to implement one.

Reply:
I didn't want to put `add()` implementation into this PR but it makes sense to implement `add()` method and change the logic to `currentHasContainerHolder.getHasContainers().forEach(...)` but I would prefer to do that in a separate PR. We already used `stream` here

Reply:
Alright, thanks for clarification!